[WIP] Add first pass at panther++

mrecachinas opened the pull request:
<!--
Please run black to format your code.
See https://networkx.org/documentation/latest/developer/contribute.html for details.
-->

As discussed, follow on to #3886.

- Need to figure out how to convert `similarities` (i.e., `1 / || theta(v_i) - theta(v_j) ||`) into a true similarity metric. Perhaps just normalized to between 0 and 1?
- Probably should pull out the common core of `panther_similarity` and `panther_vector_similarity`
- Rename `panther_similarity` to `panther_path_similarity`?

rossbar commented:
@mrecachinas thanks for this PR and #3886!

There will be a NetworkX release in the near future (O[weeks]) which will have `panther_similarity` in it. One thing we would like to try to confirm is that the API for Panther and Panther++ is consistent, and that future development in this PR won't necessitate changes in the `panther_similarity` API. For example, in one of the bullets above, you mention changing the name from `panther_similarity` to `panther_path_similarity`. Changing the function name/signature for `panther_similarity` is painless now, but will require a deprecation cycle after release. I just wanted to check to see how happy you were with the current function names and signatures. Ideally we'd be able to get the function names and signatures settled prior to the NX 2.6 release. LMK what you think!

mrecachinas commented:
@rossbar I'm good with either -- definitely up to you all. The paper refers to the actual Panther path similarity algorithm simply as "Panther" and the vector similarity algorithm as "Panther++" -- 

![image](https://user-images.githubusercontent.com/924243/108584818-3cb48080-7312-11eb-840d-7f1c6aa576ec.png)

I wasn't sure about naming it `panther_plus_plus` (since can't do `panther++`), but I'm happy with whichever you prefer.

storopoli commented:
Guys, since I was the one who brought the panther paper as a feature request, I will chip in and give my two cents.

Since the algorithm and code returns a dictionary of nodes, I prefer `panther_similarity` without any `paths`. We are not returning paths I guess only similar nodes.

Also `panther_similarity_plus_plus` is a little big. What about `panther_similarity_pp`?

dschult commented:
+1 for the `_pp` suffix. That's what VF2++ will use: either `vf2pp` or `vf2_pp`

mrecachinas commented:
Sounds good!

mrecachinas commented:
Also, wanted to revisit if @dschult @rossbar @jarrodmillman had anymore thoughts on a more permanent home for `generate_random_paths` (i.e., a Markovian random walk)?

dschult commented:
I think we should leave `generate_random_paths` where it is until NXEP3 is figured out. Then there will be many more "path constructors" to put this with.  (The idea is that many of our graph generators are really just constructing edges and/or paths so maybe we should make those available as edge constructors rather than -- or in addition to -- graph constructors.)

rossbar commented:
Re: naming, I agree with @dschult and @storopoli : **Panther** -> `panther_similarity` and **Panther++** -> `panther_similarity_pp`.

> I think we should leave generate_random_paths where it is until NXEP3 is figured out. 

I agree. If it is likely that `generate_random_paths` will be moved or wrapped up in some other object, it might be worth making it "private" for the 2.6 release (i.e. prepend `_` to the name). That way, it will still be available but the `_` will indicate to users that there may be a name change/API reshuffle on the horizon. OTOH this may be overly cautious, and the arguments for moving it somewhere else may never come to anything. I'm ambivalent, but if others think that there is a very high likelihood that we'd want to move `generate_random_paths` to a new location in the near future, it's worth thinking about.

jarrodmillman commented:
I would go with `_generate_random_paths` for the 2.6 release, and revisit it with NXEP3.  The rename could be a separate PR, since it should be merged quickly.  It probably makes sense to add a note to NXEP3 mentioning `_generate_random_paths` in the same rename PR, so that we don't forget to revisit this.

dschult commented:
The problem with making it a private function is that people won't see it or know they can use it. This is potentially a useful function for sampling paths. We shouldn't hide it.  If we use an underscore and still put it in the `__all__` variable and have it in the docs and in the main namespace, then I'm fine with making it private. But I'd prefer not to hide it from the documentation.

If it was up to me, I would leave it where it is without an underscore. If we get NXEP3 in place for v3.0 then I feel like we can still move the function to a new location within the package for v3.0. The API for the function will be the same. The function will still be the same -- we're only deprecating the location within the package. `nx.generate_random_paths` will work exactly as it does now.  Only users who used the full path to the function will have to change, and the error they get will be clear and easy to fix.

I'm open to adding the `_` if that's what should be done -- I'm just wanting verification that changing the location of a function within the package is a big enough change that the cost of keeping this function private is worth the benefit.

jarrodmillman commented:
I think we all understand the arguments for making it public or private.  So I will leave it to @dschult to decide.  It if is made public in 2.6 and deprecated in 2.7, we can still remove it in 3.0.  If it is deprecated after the 3.0 release, it can be removed in 3.2.

jarrodmillman commented:
@dschult Here is another suggestion (but you and Ross should make the call on this):

We could mark the function private in `networkx/algorithms/similarity.py`; but then import it in the top-level namespace as follows:
```
from networkx.algorithms.similarity import _generate_random_paths as generate_random_paths
```

dschult commented:
Interesting... Lots of Options...  
Maybe @rossbar and I can explore the possibilities and come up with a rationale for when to do which. :}

rossbar commented:
TL;DR - I agree with @dschult , I'm +1 on having `generate_random_paths` remain fully public (no `_`-prepended) for 2.6 and beyond.

I think the points in @dschult 's points are the most compelling. One thing that I hadn't considered was that `generate_random_paths` is available in the top-level namespace (i.e. `nx.generate_random_paths`), so which pkg/module it is defined in is basically an implementation detail that isn't interesting for most *users*. It can be moved to pretty much any module, including a new one - as long as it's still imported into the top-level namespace, the usage-impact will be minimal. Of course, moving it would cause minor changes, e.g. where it can be found in the reference docs etc, but the search bar alleviates that concern.

mrecachinas commented:
Huge apologies for the ...4 year delay @dschult @rossbar @storopoli @jarrodmillman 🙇 

Finally had some time this weekend to take a look at what was going on, finish the implementation, and get it passing CI. Let me know if you have any questions/comments after you review.

dschult commented on networkx/algorithms/similarity.py:
```diff
@@ -1642,6 +1643,170 @@ def panther_similarity(
+def panther_vector_similarity(
+    G, source, D=50, k=5, path_length=5, c=0.5, delta=0.1, eps=None, weight="weight"
+):
+    r"""Returns the Panther vector similarity of nodes in the graph ``G``
+    to node ``v``. This is known as "Panther++".
+
+    Panther is a similarity metric that says
+    "two objects are considered to be similar if their structures
+    are similar" [1]_.
```
```suggestion
    r"""Returns the Panther vector similarity of nodes in `G`.

     This approach to similarity is often called "Panther++" and the notion of
     similarity referred to here is low "graph edit distance". Panther is a
     similarity metric that says "two objects are considered to be similar
     if their structures are similar" [1]_. In this case, if their graph edit
     distances are small.
```

dschult commented on networkx/algorithms/similarity.py:
```diff
@@ -1642,6 +1643,170 @@ def panther_similarity(
+
+    Returns
+    -------
+    similarity : dictionary
+        Dictionary of nodes to similarity scores (as floats). Note:
+        the self-similarity (i.e., ``v``) will not be included in
+        the returned dictionary.
```
```suggestion
    similarity : dict
        Dict of nodes to similarity scores (as floats).
        Note: the self-similarity (i.e., `node`) is not included in the dict.
```

dschult commented:
Welcome back! :)

dschult commented on networkx/algorithms/similarity.py:
```diff
@@ -1642,6 +1643,170 @@ def panther_similarity(
 
+@nx._dispatchable(edge_attrs="weight")
+def panther_vector_similarity(
+    G, source, D=50, k=5, path_length=5, c=0.5, delta=0.1, eps=None, weight="weight"
```
Let's make these parameters keyword only -- since this is a new function that will lead to easier to read code using this. 
```suggestion
    G, source, *, D=50, k=5, path_length=5, c=0.5, delta=0.1, eps=None, weight="weight"
```

dschult commented on networkx/algorithms/similarity.py:
```diff
@@ -1642,6 +1643,170 @@ def panther_similarity(
+    path_length : int
+        How long the randomly generated paths should be (``T`` in [1]_)
+    c : float
+        A universal positive constant. Defaults to 0.5 per [1]_
```
Say what this constant does -- (the fact that it is constant can be said about any of these parameters).
Something like "Scale of the sample size when computing ___."

dschult commented on networkx/algorithms/similarity.py:
```diff
@@ -1642,6 +1643,170 @@ def panther_similarity(
+        A universal positive constant. Defaults to 0.5 per [1]_
+    delta : float
+        The probability that $S$ is not an epsilon-approximation to (R, phi)
+    eps : float
+        The error bound. Per [1]_, a good value is ``sqrt(1/|E|)``. Therefore,
+        if no value is provided, the recommended computed value will be used.
```
Need to say which error you are talking about. And "the recommended computed value" doesn't mean anything without more information. Give the exact formula if you'd like. But be more specific in any case.

dschult commented on networkx/algorithms/similarity.py:
```diff
@@ -1642,6 +1643,170 @@ def panther_similarity(
+
+    num_nodes = G.number_of_nodes()
+    if num_nodes < k:
+        warnings.warn(
+            f"Number of nodes is {num_nodes}, but requested k is {k}. "
+            "Setting k to number of nodes."
+        )
+        k = num_nodes
```
I'd prefer to raise when k > len(G) instead of trying to guess what the user wanted.
Same is true for D and maybe others.

dschult commented on networkx/algorithms/similarity.py:
```diff
@@ -1642,6 +1643,170 @@ def panther_similarity(
+        )
+        D = num_nodes
+
+    # According to [1], they empirically determined
+    # a good value for ``eps`` to be sqrt( 1 / |E| )
```
```suggestion
    # In [1], an empirically good value for ``eps`` is sqrt( 1 / |E| )
```

dschult commented on networkx/algorithms/similarity.py:
```diff
@@ -1642,6 +1643,170 @@ def panther_similarity(
+    t_choose_2 = math.comb(path_length, 2)
+    sample_size = int((c / eps**2) * (np.log2(t_choose_2) + 1 + np.log(1 / delta)))
+    index_map = {}
+    _ = list(
+        generate_random_paths(
+            G, sample_size, path_length=path_length, index_map=index_map, weight=weight
+        )
+    )
```
If the output here is not used, you don't need to assign it. Just call `generate_random_paths`. Thought either way it might help to have a comment soemthing like: `# calling to fill index_map`

dschult commented on networkx/algorithms/tests/test_similarity.py:
```diff
@@ -949,3 +949,136 @@ def user_match(x, y):
+        G.add_edge(1, 2)
+        G.add_edge(2, 4)
+
+        # Just verify that the function works and returns results
```
```suggestion
```

dschult commented on networkx/algorithms/tests/test_similarity.py:
```diff
@@ -949,3 +949,136 @@ def user_match(x, y):
+
+        sim = nx.panther_vector_similarity(G, 0, D=3, k=4, path_length=2)
+
+        # Check that we get the expected number of results
```
Don't need these comments -- the python code itself is almost english. :)

dschult reviewed (commented):
Here are some comments after a pass through this PR. I am still putting together big picture perspectives... but it looks good.

Can you have the docs explain why the function name is what it is? The docs don't talk about vector. Also how does it differ from panther_sim...? Why/when might you want to use this?

mrecachinas commented:
@dschult I think I added changes for each of your review comments. Let me know if that looks more in line with what you were thinking.

Also, let me know if I should make similar adjustments in `panther` to the docstring and change the warnings to exceptions. This feels like an API-breaking change, so I'm happy either way.

dschult commented:
Thanks for this question -- I hadn't opened up the file wider to see the similarity between panther and panther++.  MNost of that code is the same -- checking inputs and computing the panther path stats.

To help us avoid divergence in those code-paths does it make sense to have a helper function that computes as much as it can that is common?  Then if someone fixes an issue for e.g. panther, it will get fixed for panther++ too without having to remember to change it in both places. You can make that function "private" by starting the name with an underscore `_` and  dont put it in `__all__` at the top. Also don't worry about docs more than comments to help code-readers understand what is going on. 

I think that code change will also help me easily see what is different between similarities and vector similarities. The panther path generation will be the same, and the processing of the paths will hold the key to knowing the difference.

And -- yes -- let's update the panther code to match the panther++.  I'll bring up your questions about backward compatibility at the community meeting. The change will not lead to a hidden (surprise!) change -- it will raise an exception with a hopefully helpful message so it will be clear how to fix their code. That is, of course a change -- but better than a potentially hidden change.

dschult reviewed (approved):
This looks good to me!

amcandio commented on networkx/algorithms/similarity.py:
```diff
@@ -1505,6 +1507,124 @@ def _simrank_similarity_numpy(
+    index_map = {}
+
+    # Generate the random paths and populate the index_map
+    list(
+        generate_random_paths(
+            G, sample_size, path_length=path_length, index_map=index_map, weight=weight
+        )
+    )
```
Do you need to create the whole list here? Might be better to do something like
```suggestion
        for _ in generate_random_paths( 
             G, sample_size, path_length=path_length, index_map=index_map, weight=weight
        ):
            pass
```

mrecachinas replied:
Good point.

amcandio reviewed (approved):
Very cool job! Left some minor comments / suggestions but this look good to me

amcandio commented on networkx/algorithms/similarity.py:
```diff
@@ -1642,6 +1751,168 @@ def panther_similarity(
+
+        # Build up the feature vector using the
+        # largest ``D`` similarity scores
+        theta[vi_idx] = np.sort(np.partition(S[vi_idx], -D)[-D:])[::-1]
```
Do you really need this matrix S? I see you use it to compute the top D scores. What about creating a new heap per node in the for loop and keeping the top D there?

mrecachinas replied:
Good idea. Take a look at https://github.com/networkx/networkx/pull/4400/commits/af878a2dc084e842b2e0e5f658ebee485f1b6107.

Idle thoughts:
- this should increase memory efficiency -- before I was calculating the NxN matrix (`O(n^2)` memory), but now it's just the top `D` scores (`O(n*D)`)
- time complexity will be the same, but we'll be avoiding sorting on each run

mrecachinas commented on networkx/algorithms/similarity.py:
```diff
@@ -1505,6 +1507,124 @@ def _simrank_similarity_numpy(
+        node_map=list(G),
+        sample_size=sample_size,
+        index_map=index_map,
+        inv_sample_size=1 / sample_size,
```
So `PantherPaths` is more of a way to return the results from `_prepare_panther_paths` in a NamedTuple (i.e., without just returning 6-7 variables and worrying about any weirdness with unpacking/indexing). Also, `PantherPaths`, along with `_prepare_panther_paths` is scoped to this module and, more importantly not exposed outside of it.

To answer your question though: I don't think it makes sense for it to be 0 given that would blow up `inv_sample_size`, which is used later on, which would blow up the similarity metrics.

mrecachinas commented on networkx/algorithms/similarity.py:
```diff
@@ -1505,6 +1507,123 @@ def _simrank_similarity_numpy(
+        "node_map",
+        "sample_size",
+        "index_map",
+        "inv_sample_size",
```
Good catch!

mrecachinas commented on networkx/algorithms/similarity.py:
```diff
@@ -1642,6 +1751,168 @@ def panther_similarity(
+            # Only consider paths where both
+            # ``node`` and ``source`` are present
+            common_paths = set(index_map[vi]).intersection(paths)
+            S[vi_idx, inv_node_map[node]] = len(common_paths) * inv_sample_size
```
No it's fair feedback to avoid recomputing the set in each inner iteration. I'm not sure what I'd sum over since the intersection will return a result though.

amcandio replied:
The sum would save you the intersection set creation and mem allocation. I meant sum(1 for x in set1 if x in set2) (sorry for bad formatting, reviewing from my phone 😅)

This is only significant if the intersection is big enough, otherwise it's just a micro-optimization. Intersection+len is nicer/clearer

mrecachinas commented:
Not sure why test/extra is failing, but it looks unrelated?

amcandio commented on networkx/algorithms/similarity.py:
```diff
@@ -1642,6 +1749,187 @@ def panther_similarity(
+        vi_paths = index_map_sets[vi]
+
+        # Use a min heap to track the top D scores
+        # Since heapq is a min heap, we negate scores to track largest values
```
I think you want to pop the minimum, right? So you shouldn't negate values

mrecachinas replied:
No, higher similarity score == more similar, so we want a max heap here.

amcandio replied:
When you do pop, do you want to pop the minimum score or the maximum score?

amcandio replied:
Please bear with my confusion 😅 If you pop the maximum it means that you end up with the smallest scores in your heap. Is that what you are trying to do?

amcandio replied:
So let's say your scores are 1,2,3,4,5 and you want the highest 3.

With a min heap (without negating) you will pop 1 and 2, ending up with 3, 4 and 5.

With a max heap (negating) you will pop -4 and -5, ending up with 1, 2 and 3.

So if you want the highest scores, you shouldn't negate. Does this make sense?

mrecachinas replied:
Ah! I misunderstood and thought you meant when we go to pop the N largest off. Yep, I had the `elif` branch wrong where it was replacing the root node in the "max-heap".

Swapped it out to be a "min-heap" and am no longer negating.

amcandio commented:
Looks good to me, no more comments from my side!

dschult commented:
Can you merge main into this branch -- there is a fix @rossbar put in to get rid of some `extra` errors. And merging that into this will tell us if there is anything else affected there. I doubt there is anything after the found error, but it's good to make sure. :)

Thanks!!

Peiffap commented on networkx/algorithms/similarity.py:
```diff
@@ -1642,6 +1749,182 @@ def panther_similarity(
+        # Get the set of paths for the current node
+        vi_paths = index_map_sets[vi]
+
+        # Use a min heap to track the top D scores
```
I think this can be made faster using quickselect/introselect, but I don't know if the added code complexity is worth the reduction in time complexity (O(n + D log D) instead of O(n log D)). What are typical values of n and D?

Peiffap replied:
This is out of scope for this PR and might not even be faster in practice.

mrecachinas replied:
My understanding is the typical number of nodes is 100k+ or millions, whereas k and D are usually 50-100.

Peiffap reviewed (commented):
Some nit-picky comments :)

amcandio commented:
@dschult are we good to merge or should we get more reviewers?

Peiffap commented on networkx/algorithms/tests/test_similarity.py:
```diff
@@ -982,3 +982,120 @@ def user_match(x, y):
+        G.add_edge(1, 2)
+        G.add_edge(2, 4)
+
+        sim = nx.panther_vector_similarity(G, 0, D=3, k=4, path_length=2)
```
This one doesn't use a fixed seed.

Peiffap commented on networkx/algorithms/similarity.py:
```diff
@@ -1505,6 +1507,121 @@ def _simrank_similarity_numpy(
+    index_map = {}
+
+    # Generate the random paths and populate the index_map
+    for _ in generate_random_paths(
```
This call doesn't use a seed and makes the `panther*similarity` functions non-deterministic. Do we want them to be? Would it make sense to always explicitly use the same seed here, or to let the user specify a seed when calling `panther*similarity` (using an additional argument to the function)?
I'm asking because some of the tests occasionally fail if the random state is not set explicitly beforehand. I've indicated the ones I broke below, but there might be others.

If you think it's better to fix random seeds in our tests (as opposed to exposing them to users or setting them in the code), I think setting `np.random.seed(42)` works everywhere.

dschult replied:
Yes -- let's make sure that any place that uses random number generators has a seed in the function signature (with the associated random decorator on the function) and that it passes that `seed` parameter to all internally called functions.  

Thanks @Peiffap !!

mrecachinas replied:
Think I got everything in my two most recent commits https://github.com/networkx/networkx/pull/4400/commits/4a1ee39cca75cbcfbc030341ceb44f0ccc4b6d29 and https://github.com/networkx/networkx/pull/4400/commits/79e43c68677f6db2a0d5a2cb2f5e5be194bad6da.

Peiffap commented on networkx/algorithms/tests/test_similarity.py:
```diff
@@ -982,3 +982,120 @@ def user_match(x, y):
+
+        sim = nx.panther_vector_similarity(G, 0, D=3, k=4, path_length=2)
+
+        assert len(sim) == 3
+
+        assert 0 not in sim
+
+        for node in [1, 2, 3]:
+            assert node in sim
+
+        for node, score in sim.items():
+            assert 0 <= score <= 1
```
I like how this looks in the function above
```suggestion
        assert len(sim) == 3
        assert 0 not in sim
        assert all(node in sim for node in [1, 2, 3])
        assert all(0 <= score <= 1 for score in sim.values())
```

Peiffap commented on networkx/algorithms/tests/test_similarity.py:
```diff
@@ -982,3 +982,120 @@ def user_match(x, y):
+            G, "v1", D=3, k=4, path_length=2, weight="weight"
+        )
+
+        assert len(sim) == 3
+
+        assert "v1" not in sim
+
+        for node, score in sim.items():
+            assert 0 <= score <= 1
+
+        for node in ["v2", "v3", "v4"]:
+            assert node in sim
```
Same here
```suggestion
        assert len(sim) == 3
        assert "v1" not in sim
        assert all(0 <= score <= 1 for score in sim.values())
        assert all(node in sim for node in ["v2", "v3", "v4"])
```

Peiffap commented on networkx/algorithms/tests/test_similarity.py:
```diff
@@ -982,3 +982,120 @@ def user_match(x, y):
+        G.add_edge("v2", "v3", weight=0.1)
+        G.add_edge("v3", "v5", weight=1)
+
+        sim = nx.panther_vector_similarity(
```
This one doesn't use a fixed seed.

Peiffap commented on networkx/algorithms/similarity.py:
```diff
@@ -1505,6 +1507,121 @@ def _simrank_similarity_numpy(
+        - G: The graph (possibly with isolates removed)
+        - inv_node_map: Dictionary mapping node names to indices
+        - node_map: Array or list of node names
+        - sample_size: Number of paths to generate
```
My comment wasn't very clear. This got changed during code review and is no longer part of the named tuple.
```suggestion
```

Peiffap commented on networkx/algorithms/similarity.py:
```diff
@@ -1642,6 +1749,182 @@ def panther_similarity(
+            elif similarity > top_d_heap[0]:
+                heapq.heapreplace(top_d_heap, similarity)
+
+        # Extract the top D scores in descending order
+        scores = heapq.nlargest(D, top_d_heap)
```
I checked this locally and just sorting works, no need for slices.

Peiffap reviewed (commented):
Some comments regarding randomness, mainly. Also some aesthetic suggestions that you're free to disregard. :)

Peiffap commented on networkx/algorithms/similarity.py:
```diff
@@ -1642,7 +1771,198 @@ def panther_similarity(
+                heapq.heapreplace(top_d_heap, similarity)
+
+        # Extract the top D scores in descending order
+        scores = sorted(top_d_heap)
```
In an earlier state, this used [`nlargest()`](https://docs.python.org/3/library/heapq.html#heapq.nlargest), but one of the next paragraphs in that link says to use `sorted` instead for large values of `n`:
> The latter two functions perform best for smaller values of n. For larger values, it is more efficient to use the [sorted()](https://docs.python.org/3/library/functions.html#sorted) function. Also, when n==1, it is more efficient to use the built-in [min()](https://docs.python.org/3/library/functions.html#min) and [max()](https://docs.python.org/3/library/functions.html#max) functions. If repeated usage of these functions is required, consider turning the iterable into an actual heap.

As for why not pop from the heap—you could, but you end up with `lg n` complexity for each `heappop()` so it will very likely end up slower in practice than the optimized implementation in `sorted()`. And you have to reverse it...

Which does make me ask (and it's really my bad potentially since I suggested this change): shouldn't it be `sorted(top_d_heap, reverse=True)`, since we want it in decreasing order?

mrecachinas commented on networkx/algorithms/similarity.py:
```diff
@@ -1642,7 +1771,198 @@ def panther_similarity(
+    from scipy.spatial import cKDTree as KDTree
+
+    # Use helper method to prepare common data structures but keep isolates in the graph
+    paths = _prepare_panther_paths(
+        G,
+        source,
+        path_length=path_length,
+        c=c,
+        delta=delta,
+        eps=eps,
+        weight=weight,
+        remove_isolates=False,
+        k=k,
+        seed=seed,
+    )
+
+    G = paths.G
+    inv_node_map = paths.inv_node_map
+    node_map = paths.node_map
+    index_map = paths.index_map
+    inv_sample_size = paths.inv_sample_size
+    eps = paths.eps
+    num_nodes = G.number_of_nodes()
```
No major reason, just a bit of an organizational thing to avoid messing up indexing (e.g., if anything was added or removed). Happy to change it to a non-named tuple.

mrecachinas replied:
Addressed in [c06951b](https://github.com/networkx/networkx/pull/4400/commits/c06951b99323f55a95905329c499fc764b176fc6).

mrecachinas commented:
@dschult I mentioned this in a comment buried somewhere (https://github.com/networkx/networkx/pull/4400#discussion_r2094301463):

> My understanding is the typical number of nodes is 100k+ or millions, whereas k and D are usually 50-100. At least, the paper benchmarks it on hundreds of millions of nodes IIRC.

(Note too that in the original paper's benchmarks, Panther was faster than Panther++, especially for large graphs, which makes sense since it's a more "naive" calculation than building feature vectors for each node.)

I can certainly revert it back to the numpy sorting/slicing. There was also mention of quickselect, which is an option. I suspect leaning on numpy might be faster here though.

dschult commented:
I suggest using numpy -- and it looks like it doesn't matter what we use at this part of the algorithm.

dschult commented:
Can we stop optimizing the sort -- at least until we make sure that is the slow part of this algorithm?  The code we've got (with the quicksort) will never run for tens of thousands of nodes, let alone a million. 

Optimization should first identify where the bottlenecks are.

Peiffap commented:
Let me start off by saying I agree with the you that we shouldn't optimize things before we've identified them as bottlenecks. With that being said, we're in the innermost loop of the code, computing similarity between each node and _all other nodes_—and Michael claims that the algorithm is regularly used for hundreds of thousands or even millions of nodes. I'd be very surprised if that's not a significant part of the runtime... But it's not impossible of course. I'm not sure how much time the `KDTree` creation and querying will take.

Perhaps I should phrase it as a question instead: why do you say the code we have (with quicksort) will never run for tens of thousands of nodes?

It's my understanding that if we just sort and slice, we're sorting an array with `N` (the number of nodes, i.e. up to at least 1000000) entries, then slicing the top `D`, with `D` being ~100.
The heap implementation that's currently being reviewed goes through the `N` entries, keeping the top `D` and sorting only those. If that's what you're referring to, are you saying we should disregard your [original comment](https://github.com/networkx/networkx/pull/4400#issuecomment-2933025276) on the heap being slower than sorting and slicing?

Either way, the only satisfactory answer here is going to involve some more benchmarking, even if just to see whether collecting the top `D` entries is a bottleneck.

mrecachinas commented:
I just wrote a quick benchmarking script for the numpy implementation of Panther++ against a Barabasi-Albert graph. Below are the results:

(Quick note: I'm on a 2021 M1 Max MBP with 64 GB of memory.)

```
============================
Size     Nodes    Edges    Time (s)     Results 
--------------------------------------------------
10       10       21       0.2949       4       
100      100      291      0.0497       4       
1000     1000     2991     1.5536       4       
5000     5000     14991    29.7625      4       
10000    10000    29991    118.6974     4
```

Here's the core script logic with params:

```python
    for size in sizes:
        try:
            G = nx.barabasi_albert_graph(size, min(3, size-1), seed=42)
            degrees = dict(G.degree())
            source = max(degrees, key=degrees.get)
            D = min(50, G.number_of_nodes() - 1)
            panther_params = {
                'D': D,
                'k': min(5, G.number_of_nodes() - 1),
                'path_length': 5,
                'c': 0.5,
                'seed': 42
            }
```

Based on this, that probably means 100k nodes would be ~4 hours and 1M nodes would be ~10 days. It's still not as fast as the implementation in the paper. 🤔

![Panther paper benchmarks](https://github.com/user-attachments/assets/1cea21a0-5e3f-45a1-82ff-e2bb8ec2d4a9)

Peiffap commented:
What do you call the numpy implementation? What do the timings look like for `panther_similarity`? Do those match the timings in the paper?

amcandio commented:
> Another note: From some quick profiling, the two most expensive operations were:
> - `generate_random_paths` (~38%)
> - calculating path similarities (the loop) (~62%)
> 
> On a 10k node graph, this looks roughly like:
> - 52 seconds for `generate_random_paths`
> - 85 seconds for the path similarity loop
> 
> Trying something out to see if we can optimize generate_random_paths since it's the common denominator between both algorithms.

Is your graph sparse? I see generate random path creates an adjacency matrix, which might be inefficient for sparse graphs.

mrecachinas commented:
Yeah -- I was still using a BA graph.

Anyway, I'm happy to pursue further optimizations in this PR or ship the PR and pursue them in a different PR. I'll defer to you all on that.

dschult reviewed (approved):
This seems like a good approach -- too bad it isn't faster, but we can work on that. Presumably the publication wasn't using python. :}

I approve this PR!   Sorry it took so long.

rossbar reviewed (commented):
Thanks @mrecachinas , overall looks good but still some things to figure out I think.

Also when trying this out, I noticed there may be a small bug related to `k`. If I do:

```python
>>> G = nx.star_graph(100)
>>> nx.panther_vector_similarity(G, source=1, k=4, seed=42)
{26: 1.0, 52: 1.0, 98: 1.0}
```

Note that `k=4`, so I expect the result to contain the 4 most similar nodes, but there are only 3. This is true for the default to (`k=5`, but only 4 values returned). I suspect it's related to failing to account for the removal of the self-similar node from the result.

mrecachinas commented:
@rossbar Thanks for the feedback:

> Let's remove the addition of the exception for isolated nodes inside generate_random_path. Off the top of my head there are two relatively simple alternatives that don't affect backward compatibility:

Yeah good call here. Should be addressed in cb5cf49c82ece25cff5670e92ec46336cc975a74.

> The doctest failure shows that the result is not deterministic. That may be okay, but let's double-check the source of the non-determinism and ensure it's not a result of failing to seed something.

Struggling with this one -- I can't reproduce the error in CI. 🤔 

> The other test failure also seems real - I'd guess whatever fix was put in place for the k-off-by-one issue isn't working properly for k=1.

The issue was in querying the KD-tree. There was an implicit assumption that the source node would always be returned in the `k` requested nodes, hence why I updated it to query for `k+1`. Unfortunately, for a star graph, there was no guarantee the source would be returned. Should be fixed in [cccb49f](https://github.com/networkx/networkx/pull/4400/commits/cccb49f3041ecc6369c546c5d164b66e2a195c8e). Basically, I added a check if we had `k` results, and if not, took the top `k` that were returned (sorted by similarity).

mrecachinas commented:
@rossbar I suspect the nondeterminism may be coming from "ties" when querying the KD-tree. I'm sorting _after_ the ties, but I don't know how the KD-tree internally would handle ties/stability.

What are your thoughts on how to proceed here? I could modify the doctest to simply show all of the returned similarities are 1?

rossbar commented:
> I suspect the nondeterminism may be coming from "ties" when querying the KD-tree. I'm sorting after the ties, but I don't know how the KD-tree internally would handle ties/stability.

I see - in that case it's probably worth a note in the docstring.

On a related note - while reading through the [kdtree docstring](https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.KDTree.html) I noticed this:

> For large dimensions (20 is already large) do not expect this to run significantly faster than brute force. High-dimensional nearest-neighbor queries are a substantial open problem in computer science.

This seems directly relevant to our default choice of `D`, which is 50. IIUC, we're essentially constructing the tree out of an NxD array, which would be very inefficient. I see that the default `D=50` appears to have come from the paper, but perhaps we should consider a more reasonable (i.e. smaller) default given computational limitations. Note that KDTree switches over to brute force via the `leafsize` parameter, with a default value of 10. Again, IIUC, this means we're essentially getting no benefit from similarity vectors in this way.

mrecachinas commented:
@rossbar Think I covered your last two points. Lmk.

mrecachinas commented:
@dschult @rossbar Let me know if I've covered everything or if there's any remaining items.

mrecachinas commented:
@dschult I've changed it to 10.

Thinking about this a bit more, the `leafsize` kwarg is "the number of points at which the algorithm switches over to brute-force", and it defaults to 10. Right now we just use the default. They also say going above 20 won't be _significantly faster_ than brute force. I'm wondering what the metrics are in practice and whether we should set leafsize based on `D` if it's still generally, although not significantly, faster than brute force.

In any case, I did change it from 15 -> 10.

dschult commented:
The sklearn version of KDtree has a default dimension of 40. 
And there is a pure python version of KDtree that claims to be about 60 lines of code.
But it has a CC0 license and I'm not sure if that plays nicely with us, but it means it probably wouldn't be too hard to code it up. 

I'm thinking we should probably merge this and figure out how to handle the KDtree issues in follow-up PRs. Thoughts?

mrecachinas commented:
SGTM
